bsp/src/org/jetbrains/bsp/BSPCli.scala

```diff
@@ -0,0 +1,132 @@
+package org.jetbrains.bsp
+
+import java.io.File
+import java.util
+import java.util.UUID
+import java.util.concurrent.CompletableFuture
+
+import ch.epfl.scala.bsp4j
+import ch.epfl.scala.bsp4j._
+import com.intellij.openapi.externalSystem.model.ProjectSystemId
+import com.intellij.openapi.externalSystem.model.task.{ExternalSystemTaskId, ExternalSystemTaskNotificationEvent, ExternalSystemTaskNotificationListener, ExternalSystemTaskType}
+import org.jetbrains.bsp.project.resolver.BspProjectResolver
+import org.jetbrains.bsp.protocol.BspCommunication
+import org.jetbrains.bsp.protocol.BspNotifications.BspNotification
+import org.jetbrains.bsp.protocol.session.BspSession.{BspServer, BspSessionTask}
+import org.jetbrains.bsp.settings.BspExecutionSettings
+
+import scala.collection.JavaConverters._
+import scala.concurrent.Await
+import scala.concurrent.duration._
+import scala.io.StdIn
+import scala.language.postfixOps
+
+/**
+ * To run the program enter the following command in sbt interactive shell
+ * run-main org.jetbrains.bsp.protocol.session.BSPClientCliRunner /path/to/project
+ * To set a custom path to the trace log of BSP pass an env variable BSP_TRACE_PATH=/path/to/project
+ *
+ * Supported commands are
+ * - compile
+ * - test
+ * - exit
+ *
+ * compile and test are called against previously resolved build target URIs.
+ */
+object BSPCli extends App {
+
+  class DummyListener extends ExternalSystemTaskNotificationListener {
```

Comment:
I wonder if I can easily implement a trait with all methods having empty bodies ?

Reply:
This is about as easy as it gets :)